source/game/system/map/MapdataCheckPoint.cc

```diff
@@ -0,0 +1,258 @@
+#include "MapdataCheckPoint.hh"
+#include <Common.hh>
+#include <egg/math/Vector.hh>
+#include <game/system/CourseMap.hh>
+
+namespace System {
+
+MapdataCheckPoint::MapdataCheckPoint(const SData *data)
+    : m_rawData(data), m_nextCount(0), m_prevCount(0) {
+    u8 *unsafeData = reinterpret_cast<u8 *>(const_cast<SData *>(data));
+    EGG::RamStream stream = EGG::RamStream(unsafeData, sizeof(SData));
+    read(stream);
+    m_midpoint = EGG::Vector2f((left().x + right().x) / 2.0f, (left().y + right().y) / 2.0f);
+    m_dir = EGG::Vector2f(left().y - right().y, right().x - left().x);
+    m_dir.normalise();
+}
+
+void MapdataCheckPoint::read(EGG::Stream &stream) {
+    m_left.read(stream);
+    m_right.read(stream);
+    m_jugemIndex = stream.read_u8();
+    m_type = stream.read_s8();
+    m_prevPt = stream.read_u8();
+    m_nextPt = stream.read_u8();
+}
+
+/// @addr{0x80510D7C}
+/// @see MapdataCheckPoint::checkSectorAndCheckpointCompletion_
+MapdataCheckPoint::SectorOccupancy MapdataCheckPoint::checkSectorAndCheckpointCompletion(
+        const EGG::Vector3f &pos, float *completion) const {
+    EGG::Vector2f p1 = right();
+    p1.y = pos.z - p1.y;
+    p1.x = pos.x - p1.x;
+
+    for (s32 nextIdx = 0; nextIdx < m_nextCount; nextIdx++) {
+        EGG::Vector2f p0 = m_nextPoints[nextIdx].checkpoint->left();
+        p0.y = pos.z - p0.y;
+        p0.x = pos.x - p0.x;
+        MapdataCheckPoint::SectorOccupancy result =
+                checkSectorAndCheckpointCompletion_(m_nextPoints[nextIdx], p0, p1, completion);
```

Comment:
You use `m_nextPoints` in quite a few places, yet you never assign anything to this array

Reply:
there's [a big function `MapdataCheckPoint::initCheckpointLinks`](https://github.com/KooShnoo/Kinoko/blob/checkpoints/source/game/system/map/MapdataCheckPoint.cc#L28-L89) which i left out of scope of this PR which sets `m_nextPoints` and prevpts.

Reply:
If the setting of `m_nextPoints` is out-of-scope, then I feel like the getting of `m_nextPoints` should also be out-of-scope. Could you cut your changes from this PR and instead add them to the next PR you have in mind? Because right now this just feels like arbitrary picking-and-choosing of what is "close enough" to being in-scope

Reply:
you are right, i just arbitrarily cut things until PR looked small enough 😅 
should i leave just this function out then, or should i also cut the functions it calls, e.g `checkSectorAndCheckpointCompletion_`, `checkSector`, `checkCheckpointCompletion`

Reply:
If the functions you listed are only ever called from out-of-scope functions, they too should be removed